### ISSUE TITLE
SVG file displaying error - "clipPath" element problem

### ISSUE THREAD

**Issue by lukaszkrystek** [SVGKit/SVGKit#591]
### New Issue Checklist

* [+] I have read and understood the [CONTRIBUTING guide](https://github.com/SVGKit/SVGKit/blob/2.x/CONTRIBUTING.md)
* [+] I have reproduced the problem while running in Xcode, and read the Console log, and acted upon any relevant error or warning messages
* [+] I have reproduced the problem while running in Xcode's Debugger, and used breakpoints to examine the actual values of variables vs the expected values
* [+] I am not using CocoaPods, or: I am using CocoaPods, but have verified my podspec points to SVGKit v2.x

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

Some SVG files don't display properly.
They display properly in browsers (Chrome, Safari). However, using SVGKit
(iOS/macOS) they do not display correctly.
Please see attached files.

[SVG_not_supported.zip](https://github.com/SVGKit/SVGKit/files/2515010/SVG_not_supported.zip)

**Comment by adamgit** [comment]
Thanks. Can you be specific about which features aren't displaying?

The problem here is that if you don't say WHAT isn't rendering, no-one
will find it when searching for particular features / issues they want
to work on.

And also: no-one has any idea if it's an area they want to fix.

**Comment by lukaszkrystek** [comment]
Ok, It's my error - only one attached file does not display - icon_invite.svg
Other looks okey in the newest version of SVGKit.

I've attached screenshots, how it looks like in browsers (Chrome, Safari) and how it looks like in Demo-iOS application.

[screenshots.zip](https://github.com/SVGKit/SVGKit/files/2515217/screenshots.zip)

Thanks in advance.

**Comment by adamgit** [comment]
If you attach the files individually, it's more likely someone will
look into it.

If you spend some effort in looking at your SVG files and working out
which path/command/SVG tag is not rendering correvtly, it's about ten
times more likely someone might try to fix it.

**Comment by lukaszkrystek** [comment]
OK thanks - I will try to find which tag does not render correctly. Also I attached screenshots individually.

<img width="1227" alt="chrome - ok" src="https://user-images.githubusercontent.com/13151687/47508095-bee02780-d873-11e8-979c-0d388c358e78.png">

<img width="840" alt="svgkit ipad - error" src="https://user-images.githubusercontent.com/13151687/47508105-c273ae80-d873-11e8-87b0-ff6887aedfc3.png">

**Comment by lukaszkrystek** [comment]
I don't know why attached icon_invite.svg doesn't work. It contains tags like svg, path, line, clipPath, rect, g, defs, style. I'm not sure, but I think they are supported in SVGKit - I see these tags in source code.

Warning in logs: `2018-10-25 19:10:35:444 Demo-iOS[7724:453674] [SVGKParseResult] SVG WARNING (recoverable): Error Domain=SVGKit Code=1 "<clipPath> found in SVG. May render incorrectly with SVGKFastImageView due to Apple bug in CALayer .mask rendering." UserInfo={NSLocalizedDescription=<clipPath> found in SVG. May render incorrectly with SVGKFastImageView due to Apple bug in CALayer .mask rendering.}`

Rendering doesn't work both in SVGKLayeredImageView and in SVGKFastImageView. 

I hope somebody can help with this file.

Thanks

[icon_invite.svg.zip](https://github.com/SVGKit/SVGKit/files/2515976/icon_invite.svg.zip)

**Comment by dreampiggy** [comment]
@lukaszkrystek 
It seems the current `<clipPath>` implementations contains bug...Maybe we can fix it in the future.

Any PR is welcomed, and we're hoping some expert people (who have experience on Core Graphics, SVG, CSS, WebKit render engine) can help us to implements this.

I can also have a try by check the SVG spec and reference the [IJSVG](https://github.com/IconJar/IJSVG) this SVG framework. Which works for this image (though it's a SVG framework for macOS only).

<img width="712" alt="2018-10-30 12 09 38" src="https://user-images.githubusercontent.com/6919743/47663604-59c06500-dbd8-11e8-93d3-829c92b28b47.png">

**Comment by lukaszkrystek** [comment]
@dreampiggy And also [Macaw](https://github.com/exyte/Macaw) library (newest version) renders properly this SVG file.

**Comment by dreampiggy** [comment]
@lukaszkrystek [Maccaw](https://github.com/exyte/Macaw) seems a great library for SVG rendering. Does they perform the best compatible for modern SVG files ? Like Gradient, Pattern, Filter, Text , Animation comapred to SVGKit ?

It seems that we have no expert experienced contributor for SVG, maybe I'd love them to help.

**Comment by explaineverything** [comment]
I think, It's one of the best library for SVG files. I think, that they also have animations, gradient etc. But I am not sure about pattern, text, filter etc. But I would guess, that yes.

**Comment by lukaszkrystek** [comment]
I think, It's one of the best library for SVG files. I think, that they also have animations, gradient etc. But I am not sure about pattern, text, filter etc. But I would guess, that yes.

**Comment by dreampiggy** [comment]
After my test, it seems [Macaw](
https://github.com/exyte/Macaw) implement some of common SVG spec correctly (Which means, if they say support `clipPath`, it should works for all cases). But however, some of other feature which get implements in SVGKit, like radial gradient, text alignment, still lack of support. See their full list of W3C test cases: https://github.com/exyte/Macaw/blob/master/MacawTests/w3c-test-suite.md

I agree that SVG is too complicated for most of SDK author to implement it. I don't know why Apple has WebKit team could not port a native SVG support for iOS / macOS... 😓 

For most of cases, I prefer to use PDF image, which is supported natively in macOS && iOS (from iOS 11+), and works seamlessly using the Cocoa/Cocoa Touch `UIImageView/NSImageView`. SVG seems not a good solution for vector image rendering on mobile system.

**Comment by adamgit** [comment]
SVG is lighter weight than PDF (PDF is much more processor intensive to render).

Also ... SVG is much better at encoding rich information - e.g. I've worked on several projects where we merged XHtml docs generated by our own webserver with SVG graphics into a single SVG and rendered it but kept the metadata intact. e.g. I've seen a lot of game developers use SVG as a 2D level format (you can easily annotate shapes as "ledges" that the player can jump on, or as "damaging" that kill the player, etc).

It is very strange that Apple never supported SVG for apps!

**Comment by lukaszkrystek** [comment]
Yes, it is very strange that Apple doesn't provide any framework on iOS/macOS, that renders SVG files. Although they can be rendered in webview (WKWebView/UIWebView), it has some limitations, e.g. scaling.
I agree, that SVG is very fine format, better for rendering graphic than PDF.
If user of application on iOS/macOS wants to use SVG files instead of PDF files, I think we should give him such possibility.

**Comment by dreampiggy** [comment]
@lukaszkrystek  This issue was fixed after #601 was merged. You can also have a check.

The reason is that your SVG contains a `-` separator class name. Like `<path class="cls-3>`. However, previous we have a parser bug which treat this into two class name `cls`, `3`, it's solved by #601 

![image](https://user-images.githubusercontent.com/6919743/47998469-66920b00-e13a-11e8-8b13-738fc2906a68.png)

**Comment by lukaszkrystek** [comment]
Thanks @dreampiggy 

Which branch should I use to have these changes? 
I cannot find them.
